test/SILOptimizer/pound_assert.swift

```diff
@@ -0,0 +1,146 @@
+// RUN: %target-swift-frontend -enable-experimental-static-assert -emit-sil %s -verify -Xllvm -debug -Xllvm -debug-only -Xllvm ConstExpr
+
+//===----------------------------------------------------------------------===//
+// Basic function calls and control flow
+//===----------------------------------------------------------------------===//
+
+func isOne(_ x: Int) -> Bool {
+  return x == 1
+}
+
+func test_assertionSuccess() {
+  #assert(isOne(1))
+  #assert(isOne(1), "1 is not 1")
+}
+
+func test_assertionFailure() {
+  #assert(isOne(2)) // expected-error{{assertion failed}}
+  #assert(isOne(2), "2 is not 1") // expected-error{{2 is not 1}}
+}
+
+func test_nonConstant() {
+  #assert(isOne(Int(readLine()!)!)) // expected-error{{#assert condition not constant}}
+  #assert(isOne(Int(readLine()!)!), "input is not 1") // expected-error{{#assert condition not constant}}
+}
+
+// We don't support mutation, so the only loop we can make is infinite.
+// TODO: As soon as we support mutation, add tests with finite loops.
+func infiniteLoop() -> Int {
+  // expected-note @+2 {{condition always evaluates to true}}
+  // expected-note @+1 {{control flow loop found}}
+  while true {}
+  // expected-warning @+1 {{will never be executed}}
+  return 1
+}
+
+func test_infiniteLoop() {
+  // expected-error @+1 {{#assert condition not constant}}
+  #assert(infiniteLoop() == 1)
+}
+
+func recursive(a: Int) -> Int {
+  if a == 0 { return 0 }     // expected-note {{expression is too large to evaluate at compile-time}}
+  return recursive(a: a-1)
+}
+
+func test_recursive() {
+  // expected-error @+1 {{#assert condition not constant}}
+  #assert(recursive(a: 20000) > 42)
+}
+
+func conditional(_ x: Int) -> Int {
+  if x < 0 {
+    return 0
+  } else {
+    return x
+  }
+}
+
+func test_conditional() {
+  #assert(conditional(-5) == 0)
+  #assert(conditional(5) == 5)
+
+  // expected-error @+1 {{assertion failed}}
+  #assert(conditional(-5) == 1)
+  // expected-error @+1 {{assertion failed}}
+  #assert(conditional(5) == 1)
+}
+
+//===----------------------------------------------------------------------===//
+// Top-level evaluation
+//===----------------------------------------------------------------------===//
+
+func test_topLevelEvaluation(topLevelArgument: Int) {
+  let topLevelConst = 1
+  #assert(topLevelConst == 1)
+
+  // TODO: I think that this should get rejected. Otherwise, constness of vars
+  // depends on what you do to them, which is confusing. Perhaps some other
+  // const folder is getting to this before ConstExpr.cpp does?
+  var topLevelVar = 1 // expected-warning {{never mutated}}
```

Comment:
@marcrasi The answer to the question in the comments is Yes. The `DiagnosticConstantPropagation` phase runs before this phase and it folds this constant. This I think should be okay, it only makes #assert's a little bit more powerful ;-)